File: ios/SampleApp/Screens/Detail/ItemDetailViewModel.swift
import SwiftUI

enum ItemDetailState: Equatable {
    case loading
    case success(Item)
    case error
}

@MainActor
final class ItemDetailViewModel: ObservableObject {
    @Published private(set) var state: ItemDetailState = .loading

    private var itemId: String?

    func load(id: String) async {
        itemId = id
        state = .loading
        if let item = await MockRepository.shared.getItem(id: id) {
            state = .success(item)
        } else {
            state = .error
        }
    }

    func toggleFavorite() async {
        guard let itemId else { return }
        await MockRepository.shared.toggleFavorite(id: itemId)
        if let item = await MockRepository.shared.getItem(id: itemId) {
            state = .success(item)
        } else {
            state = .error
        }
    }
}

File: android/app/src/main/java/com/example/sampleapp/ui/detail/ItemDetailViewModel.kt
package com.example.sampleapp.ui.detail

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.example.sampleapp.data.Item
import com.example.sampleapp.data.MockRepository
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch

sealed interface ItemDetailState {
    data object Loading : ItemDetailState
    data class Success(val item: Item) : ItemDetailState
    data object Error : ItemDetailState
}

class ItemDetailViewModel : ViewModel() {
    private val _state = MutableStateFlow<ItemDetailState>(ItemDetailState.Loading)
    val state: StateFlow<ItemDetailState> = _state.asStateFlow()

    private var itemId: String? = null

    fun load(id: String) {
        if (itemId == id && _state.value is ItemDetailState.Success) return
        itemId = id
        viewModelScope.launch {
            _state.value = ItemDetailState.Loading
            _state.value = MockRepository.getItem(id)
                ?.let(ItemDetailState::Success)
                ?: ItemDetailState.Error
        }
    }

    fun toggleFavorite() {
        val id = itemId ?: return
        viewModelScope.launch {
            MockRepository.toggleFavorite(id)
            _state.value = MockRepository.getItem(id)
                ?.let(ItemDetailState::Success)
                ?: ItemDetailState.Error
        }
    }
}